### PR TITLE
Use thiserror for error handling

### PR DESCRIPTION
**Pull request by MikailBag**
# Motivation
 - `error-chain` generates Error type which is not `Sync`, so it is hard to use with many other libraries.
 - Libraries are encouraged to provide structured error details instead of just formatted message.
# Changes
 - `errors::Error` type is now plain enum, using `thiserror` for `Display` and `From` impls.
 - It is very large, but I don't think it is significant drawback.
 - Some functions return custom Error type (which is more convenient to use), but such types can be converted to `errors::Error`.
 - Unfortunately new approach does not play well with backtraces. I think this can be fixed if needed.
 - In several places I replaced `?` with `expect` because there should not be any errors in fact.

### CODE REVIEW

**Comment by lucab** [comment]
@steveeJ I think you were looking at refreshing error handling in other places too, do you maybe have some cycles to review  and land this one? I'm +1 on `thiserror` here.

**Comment by steveeJ** [comment]
> I think you were looking at refreshing error handling in other places too, do you maybe have some cycles to review and land this one? I'm +1 on `thiserror` here.

Yes, I'll be happy to take a look at this. I migrated [away from `failure` on openshift/cincinnati recently](https://github.com/openshift/cincinnati/pull/286). That PR was less invasive than this one, although I do understand the motivation of adding more specific and structured errors.

From glancing at this it looks like it'll be in conflict with #159, so I'd prefer to land that first.

**Comment by lucab** [comment]
@steveeJ ack, let's serialize these. I started going through #159 but didn't manage to finish, I'll hope to finish that on Monday.

**Comment by MikailBag** [comment]
Thanks for positive feedback!
I will be happy to rebase onto master and fix merge conflicts if needed.

**Comment by steveeJ** [comment]
@MikailBag #159 has landed. I'll review this PR when it's rebased!

**Comment by MikailBag** [comment]
I have rebased onto master; tests pass.

File: src/mediatypes.rs [hunk 1]
```diff
@@ -79,6 +79,6 @@ impl MediaTypes {
                 }
             }
         }
-        .map_err(|e| Error::from(e.to_string()))
+        .expect("to_mime should be always successful")
```
Comment (steveeJ) [inline src/mediatypes.rs:82]: I don't follow why this cannot fail. I'd prefer to leave the `Result` in place here, and let the caller make use of the `?` operator which isn't expensive.

File: src/v2/auth.rs [hunk 2]
```diff
@@ -337,7 +315,7 @@ mod tests {
         let header_value = HeaderValue::from_str(&format!(
             r#"Bearer realm="{}",service="{}",scope="{}""#,
             realm, service, scope
-        ))?;
+        )).expect("this statically known header value only contains ASCII chars so it is correct header value");
```
Comment (steveeJ) [inline src/v2/auth.rs:318]: The variables _realm_, _service_, and _scope_ aren't statically known.
Also, my rule of thumb is that if it's a 3rd party library I don't rely on it and avoid panicing.

File: src/v2/auth.rs [hunk 3]
```diff
@@ -365,7 +343,7 @@ mod tests {
     fn basic_realm_parses_correctly() -> Result<()> {
         let realm = "Registry realm";
 
-        let header_value = HeaderValue::from_str(&format!(r#"Basic realm="{}""#, realm))?;
+        let header_value = HeaderValue::from_str(&format!(r#"Basic realm="{}""#, realm)).unwrap();
```
Comment (steveeJ) [inline src/v2/auth.rs:346]: My rule of thumb is that if it's a 3rd party library I don't rely on it and avoid panicing.

File: src/mediatypes.rs [hunk 4]
```diff
@@ -62,13 +62,13 @@ impl MediaTypes {
...
-                    _ => bail!("unknown mediatype {:?}", mtype),
+                    _ => return Err(crate::Error::UnknownMimeType(mtype.clone())),
                 }
             }
-            _ => bail!("unknown mediatype {:?}", mtype),
+            _ => return Err(crate::Error::UnknownMimeType(mtype.clone())),
         }
     }
-    pub fn to_mime(&self) -> Result<mime::Mime> {
+    pub fn to_mime(&self) -> mime::Mime {
```
Comment (steveeJ) [inline src/mediatypes.rs:71]:
```suggestion
    pub fn to_mime(&self) -> Result<mime::Mime> {
```

Please see the comment below

File: src/v2/content_digest.rs [hunk 5]
```diff
@@ -1,36 +1,47 @@
-use crate::errors::Result;
 /// Implements types and methods for content verification
 use sha2::{self, Digest};
 
 /// ContentDigest stores a digest and its DigestAlgorithm
 #[derive(Clone, Debug, PartialEq)]
-pub(crate) struct ContentDigest {
+pub struct ContentDigest {
```
Comment (steveeJ) [inline src/v2/content_digest.rs:6]: Does this need to be `pub`?

File: src/v2/content_digest.rs [hunk 6]
```diff
@@ -1,36 +1,47 @@
...
     digest: String,
     algorithm: DigestAlgorithm,
 }
 
 /// DigestAlgorithm declares the supported algorithms
 #[derive(Display, Clone, Debug, PartialEq, EnumString)]
-enum DigestAlgorithm {
+pub enum DigestAlgorithm {
```
Comment (steveeJ) [inline src/v2/content_digest.rs:13]: Does this need to be `pub`?

File: src/v2/content_digest.rs [hunk 7]
```diff
@@ -1,36 +1,47 @@
...
+}
+
 impl ContentDigest {
     /// try_new attempts to parse the digest string and create a ContentDigest instance from it
     ///
     /// Success depends on
     /// - the string having a "algorithm:" prefix
     /// - the algorithm being supported by DigestAlgorithm
-    pub fn try_new(digest: String) -> Result<Self> {
+    pub fn try_new(digest: String) -> std::result::Result<Self, ContentDigestError> {
```
Comment (steveeJ) [inline src/v2/content_digest.rs:37]: What's the advantage of declaring this specific error type in this module vs. adding the variants to the crate's error module?

Comment (MikailBag) [reply src/v2/content_digest.rs:37]: If one wishes to use specifically ContestDigest, they can exhaustively match on ContestDigestError and handle all possible errors. If this function returns `crate::Error` instead, such a user would have to 
1) add wildcard check
2) write justyfing comment
3) carefully read through docs of `dkregistry::Error` (or source code) and decide for each variant if it is possible.
It is time-consuming process.

In short: IMHO, if some function returns E as error enum, and it contains variants that are never returned by this function, it is inconvenient for API user who would like to do exhaustive match.


That said, I'm ready to merge this domain-specific small enums into `crate::Error`, if you find that behavior more maintainable or readable,

File: src/v2/content_digest.rs (hunk 5 above)
Comment (MikailBag) [reply src/v2/content_digest.rs:6]: I changed visibility to `pub` because I included it into `ContestDigestError::Verify` variant details.
I *think* this can be useful for library user.
However, I see alternative option: we can include `format!`-ed digests into error instead. this way we can keep ContestDigest private.

File: src/v2/content_digest.rs (hunk 7 above)
Comment (steveeJ) [reply src/v2/content_digest.rs:37]: Thanks for elaborating, the reasoning makes sense to me! It would be nice if Rust allowed a convenient way to specify a subset of variants the of the error enum a function returns, but I can't come up with an easy way to do that. In this case I prefer the specific error type as implemented in this PR.

**Comment by MikailBag** [comment]
@steveeJ rebased.

File: src/v2/auth.rs (hunk 3 above)
Comment (steveeJ) [reply src/v2/auth.rs:352]: Bump.

File: src/v2/auth.rs (hunk 2 above)
Comment (steveeJ) [reply src/v2/auth.rs:324]: Bump.

**Review by steveeJ — CHANGES_REQUESTED** [review]
I spotted two unresolved comment threads and bumped them. I strongly prefer not to panic if we can avoid it.

**Comment by MikailBag** [comment]
Correct me if I'm wrong, but now I can see that the two mentioned lines are in tests (sorry that I have not noticed it earlier). Usually it is OK for test code to panic.
I can change `expect` to `?` in that code, but I'd like to get some confirmation from you.

File: src/v2/auth.rs (hunk 2 above)
Comment (steveeJ) [reply src/v2/auth.rs:324]: Ah, I see now that they are statically known here by expanding the previous lines. Also as you noted in your comment, this is in a test so panicking is appropriate.

**Review by steveeJ — APPROVED** [review]
> Correct me if I'm wrong, but now I can see that the two mentioned lines are in tests (sorry that I have not noticed it earlier). Usually it is OK for test code to panic.

I overlooked the fact that it's in test as well. I agree that we can stick with it there.

Thank you for your work on this!
